cfgen/cfgen

```diff
@@ -1261,6 +1270,31 @@ class ConfPool(ToDhall):
                           f'pvers = {oids_to_dhall(self.pvers)}'])
         return '{ %s }' % args
 
+    @classmethod
+    def gen_sns_tolerance(cls, m0conf: Dict[Oid, Any],
+                          pool_desc: Dict[str, Any]) -> Failures:
+        encls_nr = len(cluster_encls(m0conf))
+        ctrls_per_encl = floor(len(cluster_ctrls(m0conf)) / encls_nr)
+        data_units = pool_desc['data_units']
+        parity_units = pool_desc['parity_units']
+        spare_units = pool_desc.get('spare_units', 0)
+        total_units = data_units + parity_units + spare_units
+        enc_failure_allowed_FP = parity_units / ceil(total_units / encls_nr)
+        enc_failure_allowed = floor(enc_failure_allowed_FP)
+        temp = ctrls_per_encl * enc_failure_allowed
+        ctrl_failure_allowed = min(temp, parity_units)
+        return Failures(0, 0, enc_failure_allowed, ctrl_failure_allowed,
+                        parity_units)
+
```

Comment:
For SNS pool :
In my 6 node cluster example : 
I had 2 disks on 5 nodes and 8 disks on the 6th node. In CDF the tolerance values were not mentioned so that it will auto generate values.
It generated tolerance = [0, 0, 1, 1, 4]
Here disks = 4 and CVG =2, it means either we allow 1 CVG to fail or 4 disks to fail.
Lets assume that 2, 2 disks from first 2 nodes fail (where these nodes only had 2 nodes each).
It satisfies 4 disks allowance but its crossing the limit of 1 CVG (as what is generated automatically).

So the formula needs to be changed?

Reply:
Looks like this is for md or dix pool. Need to check this, I have some idea, probably need to get the minimum from all the nodes.

Reply:
Got input from Madhav for considering this  in case of asymmetric pools :
https://jts.seagate.com/browse/EOS-23178?focusedCommentId=2164896&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-2164896

Reply:
@suvratjoshi what were the layout parameters (N, K, S) that you used?

Reply:
N+K+S = 14+4+0

Reply:
hmm... so we are tolerating 4 failures. Now in your case with 14 + 4 + 0 and 6 nodes, 5 nodes will host 10 units of a parity group and 1 node will host remaining 4. So even if your 6th node with 8 drives fail, we will only lose maximum 4 units of a parity group, which is tolerable, same with ctrl and with 4 drives. So allowed_failures vector is fine I think.

Reply:
```
Lets assume that 2, 2 disks from first 2 nodes fail (where these nodes only had 2 disks each).
It satisfies 4 disks allowance but its crossing the limit of 1 CVG (as what is generated automatically).
```
I think tolerance disk count should be 2 in this case, so that even if 2 disks fail from a node then it will suffice all these values of allowance vector [0,0,1,1,2]

Now if we consider the pool width formula in case of asymmetric pool then it will be :
C * min(2,2,2,2,2,8) = 6 * 2 = 12

Similarly, tolerance value for disks can be taken as min(2,2,2,2,2,8), which will be 2.

Reply:
Sorry in the above reply I missed the parity units distribution, yes we need to handle asymettric pools. Have updated the PR with the related code changes. Kindly review.